src/middlewared/middlewared/plugins/pool.py

```diff
@@ -1346,6 +1347,56 @@ async def import_find(self, job):
             result.append(entry)
         return result
 
+    @private
+    async def migrate_defaults(self, ds):
+        """
+        zpool may be foreign to TrueNAS and so we need to normalize the dataset
+        properties / settings to properly work on TrueNAS. Currently, this means
+        setting some ACL settings on the root dataset, and then ensuring that
+        mountpoint is not legacy and that sharenfs and sharesmb are disabled.
+        If the operations need to be altered, we should increment the
+        TRUENAS_DEFAULTS_VERSION.
+        """
+        ds['user_properties'].append({
+            'key': 'org.truenas:truenas_defaults_version',
+            'value': str(TRUENAS_DEFAULTS_VERSION)
+        })
+        if ds['acltype'] == 'NFSV4':
+            properties = {
+                'aclinherit': 'PASSTHROUGH',
+            }
+        else:
+            properties = {
+                'aclinherit': 'DISCARD',
+                'aclmode': 'DISCARD'
+            }
+
+        properties['user_propeties'] = ds['user_proeprties']
+
+        recursive = True
+        for child in (await self.middleware.call('pool.dataset.query', [('id', '=', ds['id'])], {'get': True}))['children']:
```

Comment:
We used `zfs.dataset.child_dataset_names`, which has a better performance, hasn't it? @sonicaj Why do we need to retrieve all datasets metadata if we only need their names?

Also what about root dataset `sharenfs`, `sharesmb` and `mountpoint` properties?

Reply:
I backed out the pool changes in favor of skipping zfs_sharenfs() in py-libzfs during pool import there.

I'll see about maybe being more intelligent with how we interpret failures to inherit mountpoint property (i.e. disable sharenfs / sharesmb in this case (if they're set).